Refactor deprecated antd Dropdown menus

hotzenklotz opened the pull request:
Antd 4.2x introduced deprecation warnings for `<Menu>`components within `<Dropdown>` elements. This PR replaces the React `<Menu>` components with heir object equivalents. 

Most changes/files are rather straight forward conversions. Only the annotations context menu is more involved since is is created from a million different function, components, etc. Hopefully the strict typing of all menu items prevents any new bugs.

I tested all menus and clicked on all items but I might have missed some by mistake...

<img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExZTUzMzIzMDI1ZTliNWEwMDUzZGEzOGM3NDViYjQzMTc5MzhjOTk5ZSZjdD1n/sRIXt1e9dQRjEjxnMk/giphy-downsized.gif">

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Test that all the menu still show (or at least the most important menus)
- Consider performance. I tested with an annotation with 1700+ trees and the Trees tab remained smooth. Here is the NML file
[detection_no_size_threshold.nml.zip](https://github.com/scalableminds/webknossos/files/10957956/detection_no_size_threshold.nml.zip)
- Randomly sample some menu entries for testing

### Issues:
- Contributes to #6861 
- fixes https://github.com/scalableminds/webknossos/issues/6814

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

hotzenklotz commented on frontend/javascripts/oxalis/view/context_menu.tsx:
```diff
@@ -974,33 +985,25 @@ function NoNodeContextMenuOptions(props: NoNodeContextMenuProps): JSX.Element {
 
   if (isSkeletonToolActive) {
-    allActions = skeletonActions.concat(nonSkeletonActions).concat(boundingBoxActions);
+    allActions = [...skeletonActions, ...nonSkeletonActions, ...boundingBoxActions];
```
TS had issues with `concat()`...

hotzenklotz commented on frontend/javascripts/oxalis/view/context_menu.tsx:
```diff
@@ -1112,129 +1117,150 @@ function ContextMenuInner(propsWithInputRef: Props) {
-    const activeTreeId = skeletonTracing != null ? skeletonTracing.activeTreeId : null;
-    const activeNodeId = skeletonTracing?.activeNodeId;
+  if (contextMenuPosition == null || maybeViewport == null) {
+    return <></>;
```
Do an early return to remove one tier of nesting hierarchy.

philippotto commented on frontend/javascripts/oxalis/view/action-bar/tracing_actions_view.tsx:
```diff
@@ -164,86 +167,87 @@ export function LayoutMenu(props: LayoutMenuProps) {
+            title: "Save current layout",
+            icon: <SaveOutlined />,
+          },
+      { key: "dividier", type: "divider" },
```
```suggestion
      { key: "divider", type: "divider" },
```

hotzenklotz replied:
Done

philippotto commented on frontend/javascripts/oxalis/view/right-border-tabs/tree_hierarchy_view.tsx:
```diff
@@ -526,59 +520,65 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
         <div className={styleClass}>
           <Dropdown
-            overlay={createMenu} // The overlay is generated lazily. By default, this would make the overlay
+            menu={createMenu()} // The overlay is generated lazily. By default, this would make the overlay
```
the comment needs to be adapted (same in the other relevant modules)

philippotto commented on frontend/javascripts/oxalis/view/context_menu.tsx:
```diff
@@ -1112,129 +1117,150 @@ function ContextMenuInner(propsWithInputRef: Props) {
-    // antd's internal population of the correct class names
-    // for the menu.
-    overlay =
+  // It's important to not use <NodeContextMenuOptions ...>
```
Luckily, the comment is not needed anymore now :tada: Since these functions return JS objects instead of JSX elements, writing `<NodeContextMenuOptions ...>` wouldn't even make sense, anymore.

philippotto commented on frontend/javascripts/oxalis/view/context_menu.tsx:
```diff
@@ -467,132 +463,128 @@ function NodeContextMenuOptions({
+              },
+        ]
+      : []),
+    ...(maybeMeshItems ? maybeMeshItems : []),
```
would probably be nicer to have `getMaybeMeshItems` return an empty array instead of null so that the distinction here doesn't have to be made. then, it could also be renamed to not contain the "maybe" in the function and var name.

philippotto commented on frontend/javascripts/oxalis/view/context_menu.tsx:
```diff
@@ -1094,8 +1100,7 @@ function InfoMenuItem(propsWithEventKey: MenuItemProps) {
 
-  const { children } = propsWithEventKey;
-  return <div className="node-context-menu-item">{children}</div>;
+  return { key, label, type: "group" };
```
the above comment is not valid anymore. you could mention instead that the trick is simply to use the "group" type to get the desired behavior.

philippotto reviewed (commented):
awesome! didn't test yet, but have already left some small comments.

philippotto commented:
First round of testing:

- [ ] right clicking a dataset in the dashboard and hitting escape (or clicking somewhere else) crashes the dashboard
- [ ] right clicking a skeleton node shows a position icon which is different from the icon used in the position box. I think, using the same icon would be better
- [ ] right clicking a position in a dataset, doesn't show a position icon for me (however, there's an icon in code, I think)

philippotto commented:
One other problem:

- [ ] the `MenuItemWithMappingActivationConfirmation` component doesn't seem to work anymore (it doesn't do anything regardless of the selected mapping).

Steps to reproduce:
- visit https://dropdownmenu.webknossos.xyz/datasets/sample_organization/l4dense_motta_et_al_dev_v2/view#2878,4480,1792,0,1.3
- select meshfile_v3 in the segments tab
- rightclick a segment in the viewport and click "load mesh (precomputed)"
- now, a modal should ask you to switch to the correct mapping (instead, nothing happens)
- select agglomerate mapping 80 manually in the left sidebar
- try "load mesh (precomputed)" again --> it should load a mesh but nothing happens

philippotto commented:
I also just tested performance and everything seemed fine :+1: (for the record, I tested with 80k trees).

hotzenklotz commented:
@philippotto I think I fixed the issues that you detected and applied your PR feedback. Please have a look at my last few commits. 

I likely also fixed issue #6814 in the process.

philippotto commented on frontend/javascripts/oxalis/view/context_menu.tsx:
```diff
@@ -763,17 +756,16 @@ function NoNodeContextMenuOptions(props: NoNodeContextMenuProps): ItemType[] {
-  useEffect(() => {
-    dispatch(maybeFetchMeshFilesAction(visibleSegmentationLayer, dataset, false));
-  }, [visibleSegmentationLayer, dataset]);
+  // useEffect(() => {
```
comment can be removed now.

hotzenklotz replied:
Good catch. I mean to ask, the `useEffect(...)` call has two dependencies and will be executed when these change. I was not completely clear if the next code covers this.

hotzenklotz replied:
Done

philippotto reviewed (approved):
Splendid clean up! See only my one comment about commented code before merging.
